src/client/src/Client.c

```diff
@@ -1306,13 +1479,33 @@ STATUS freeKinesisVideoClientInternal(PKinesisVideoClient pKinesisVideoClient, S
         }
     }
 
+    // unLock the streamListLock after iteration
+    if (IS_VALID_MUTEX_VALUE(pKinesisVideoClient->base.streamListLock)) {
+        pKinesisVideoClient->clientCallbacks.unlockMutexFn(pKinesisVideoClient->clientCallbacks.customData,
+                                                         pKinesisVideoClient->base.streamListLock);
+        locked = FALSE;
+    }
+
+
     // Shutdown client
     pKinesisVideoClient->clientCallbacks.clientShutdownFn(pKinesisVideoClient->clientCallbacks.customData, TO_CLIENT_HANDLE(pKinesisVideoClient));
 
     // After the shutdown
     semaphoreLock(pKinesisVideoClient->base.shutdownSemaphore);
     semaphoreWaitUntilClear(pKinesisVideoClient->base.shutdownSemaphore, CLIENT_SHUTDOWN_SEMAPHORE_TIMEOUT);
 
+    // Need to stop scheduling callbacks

```

Comment:
Why not move this earlier and shutdown the timer queue before shutting down streams within the already held list lock?

Reply:
timerQueueShutdown won't return until callbacks are all done but the callback acquires the streams list lock, so if I move it up there we'll have a deadlock because we will never be able to acquire that so I had to be careful to make sure it's *not* within that lock.

Reply:
ouch.